studentFeedbackSubmissionEditSave: java.lang.AssertionError: null

Request Method POST
User Agent Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.101 Safari/537.36
Request Url https://teammatesv4.appspot.com/page/studentFeedbackSubmissionEditSave
Request Path /page
Request Parameters {omitted for privacy}
Stack Trace



java.lang.AssertionError: null
	at teammates.common.util.Assumption.fail(Assumption.java:59)
	at teammates.common.util.Assumption.fail(Assumption.java:66)
	at teammates.logic.core.FeedbackResponsesLogic.createFeedbackResponse(FeedbackResponsesLogic.java:53)
	at teammates.logic.api.Logic.createFeedbackResponse(Logic.java:2187)
	at teammates.ui.controller.FeedbackSubmissionEditSaveAction.saveResponse(FeedbackSubmissionEditSaveAction.java:213)
	at teammates.ui.controller.FeedbackSubmissionEditSaveAction.execute(FeedbackSubmissionEditSaveAction.java:137)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:337)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:142)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:147)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:457)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:230)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)

--- a/src/main/java/teammates/common/exception/EntityAlreadyExistsException.java
+++ b/src/main/java/teammates/common/exception/EntityAlreadyExistsException.java
@@ -2,7 +2,15 @@ package teammates.common.exception;
 
 @SuppressWarnings("serial")
 public class EntityAlreadyExistsException extends TeammatesException {
+    
+    public Object existingEntity;
+    
     public EntityAlreadyExistsException(String message) {
         super(message);
+    }
+    
+    public EntityAlreadyExistsException(String message, Object entity) {
+        this(message);
+        this.existingEntity = entity;
     }
 }

--- a/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java
@@ -421,18 +421,7 @@ public class FeedbackResponsesLogic {
                     "Trying to update a feedback response that does not exist.");
         }
 
-        copyOldResponseToNewResponse(newResponse, oldResponse);
-
-        if (!newResponse.recipientEmail.equals(oldResponse.recipientEmail) ||
-                !newResponse.giverEmail.equals(oldResponse.giverEmail)) {
-            recreateResponse(newResponse, oldResponse);
-        } else {
-            try {
-                frDb.updateFeedbackResponseOptimized(newResponse, oldResponseEntity);
-            } catch (EntityDoesNotExistException e) {
-                Assumption.fail();
-            }
-        }
+        updateFeedbackResponse(newResponse, oldResponseEntity);
     }
 
     /**
@@ -450,16 +439,35 @@ public class FeedbackResponsesLogic {
                         FeedbackResponseAttributes updatedResponse,
                         FeedbackResponse oldResponseEntity)
                                 throws InvalidParametersException, EntityAlreadyExistsException {
-        if (oldResponseEntity == null) {
-            throw new NullPointerException("oldResponseEntity should not be null");
-        }
+        Assumption.assertNotNull(oldResponseEntity);
         
         // Create a copy.
         FeedbackResponseAttributes newResponse = new FeedbackResponseAttributes(updatedResponse);
         
         FeedbackResponseAttributes oldResponse = new FeedbackResponseAttributes(oldResponseEntity);
 
-        copyOldResponseToNewResponse(newResponse, oldResponse);
+        // Copy values that cannot be changed to defensively avoid invalid
+        // parameters.
+        newResponse.courseId = oldResponse.courseId;
+        newResponse.feedbackSessionName = oldResponse.feedbackSessionName;
+        newResponse.feedbackQuestionId = oldResponse.feedbackQuestionId;
+        newResponse.feedbackQuestionType = oldResponse.feedbackQuestionType;
+        
+        if (newResponse.responseMetaData == null) {
+            newResponse.responseMetaData = oldResponse.responseMetaData;
+        }
+        if (newResponse.giverEmail == null) {
+            newResponse.giverEmail = oldResponse.giverEmail;
+        }
+        if (newResponse.recipientEmail == null) {
+            newResponse.recipientEmail = oldResponse.recipientEmail;
+        }
+        if (newResponse.giverSection == null) {
+            newResponse.giverSection = oldResponse.giverSection;
+        }
+        if (newResponse.recipientSection == null) {
+            newResponse.recipientSection = oldResponse.recipientSection;
+        }
     
         if (!newResponse.recipientEmail.equals(oldResponse.recipientEmail) 
             || !newResponse.giverEmail.equals(oldResponse.giverEmail)) {
@@ -487,32 +495,6 @@ public class FeedbackResponsesLogic {
         }
     }
 
-    
-    private void copyOldResponseToNewResponse(FeedbackResponseAttributes newResponse,
-                                    FeedbackResponseAttributes oldResponse) {
-        // Copy values that cannot be changed to defensively avoid invalid
-        // parameters.
-        newResponse.courseId = oldResponse.courseId;
-        newResponse.feedbackSessionName = oldResponse.feedbackSessionName;
-        newResponse.feedbackQuestionId = oldResponse.feedbackQuestionId;
-        newResponse.feedbackQuestionType = oldResponse.feedbackQuestionType;
-
-        if (newResponse.responseMetaData == null) {
-            newResponse.responseMetaData = oldResponse.responseMetaData;
-        }
-        if (newResponse.giverEmail == null) {
-            newResponse.giverEmail = oldResponse.giverEmail;
-        }
-        if (newResponse.recipientEmail == null) {
-            newResponse.recipientEmail = oldResponse.recipientEmail;
-        }
-        if (newResponse.giverSection == null) {
-            newResponse.giverSection = oldResponse.giverSection;
-        }
-        if (newResponse.recipientSection == null) {
-            newResponse.recipientSection = oldResponse.recipientSection;
-        }
-    }
     
     /**
      * Updates responses for a student when his team changes. This is done by

--- a/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java
@@ -46,10 +46,10 @@ public class FeedbackResponsesLogic {
     public void createFeedbackResponse(FeedbackResponseAttributes fra) throws InvalidParametersException {
         try {
             frDb.createEntity(fra);
-        } catch (Exception EntityAlreadyExistsException) {
+        } catch (EntityAlreadyExistsException eaee) {
             try {
-                updateFeedbackResponse(fra);
-            } catch (Exception EntityDoesNotExistException) {
+                updateFeedbackResponse(fra, (FeedbackResponse) eaee.existingEntity);
+            } catch (EntityAlreadyExistsException entityAlreadyExistsException) {
                 Assumption.fail();
             }
         }
@@ -421,6 +421,72 @@ public class FeedbackResponsesLogic {
                     "Trying to update a feedback response that does not exist.");
         }
 
+        copyOldResponseToNewResponse(newResponse, oldResponse);
+
+        if (!newResponse.recipientEmail.equals(oldResponse.recipientEmail) ||
+                !newResponse.giverEmail.equals(oldResponse.giverEmail)) {
+            recreateResponse(newResponse, oldResponse);
+        } else {
+            try {
+                frDb.updateFeedbackResponseOptimized(newResponse, oldResponseEntity);
+            } catch (EntityDoesNotExistException e) {
+                Assumption.fail();
+            }
+        }
+    }
+
+    /**
+     * Updates a {@link FeedbackResponse} using a {@link FeedbackResponseAttributes} <br>
+     * If the giver/recipient field is changed, the {@link FeedbackResponse} is
+     * updated by recreating the response<br>
+     * in order to prevent an id clash if the previous email is reused later on.
+     * @param updatedResponse 
+     * @param oldResponseEntity  a FeedbackResponse retrieved from the database 
+     */
+    public void updateFeedbackResponse(
+                        FeedbackResponseAttributes updatedResponse,
+                        FeedbackResponse oldResponseEntity)
+                                throws InvalidParametersException, EntityAlreadyExistsException {
+        if (oldResponseEntity == null) {
+            throw new NullPointerException("oldResponseEntity should not be null");
+        }
+        
+        // Create a copy.
+        FeedbackResponseAttributes newResponse = new FeedbackResponseAttributes(updatedResponse);
+        
+        FeedbackResponseAttributes oldResponse = new FeedbackResponseAttributes(oldResponseEntity);
+
+        copyOldResponseToNewResponse(newResponse, oldResponse);
+    
+        if (!newResponse.recipientEmail.equals(oldResponse.recipientEmail) 
+            || !newResponse.giverEmail.equals(oldResponse.giverEmail)) {
+            // Recreate response to prevent possible future id conflict.
+            recreateResponse(newResponse, oldResponse);
+        } else {
+            try {
+                frDb.updateFeedbackResponseOptimized(newResponse, oldResponseEntity);
+            } catch (EntityDoesNotExistException e) {
+                Assumption.fail();
+            }
+        }
+    }
+
+    private void recreateResponse(FeedbackResponseAttributes newResponse,
+                                    FeedbackResponseAttributes oldResponse)
+                                    throws InvalidParametersException, EntityAlreadyExistsException {
+        try {
+            newResponse.setId(null);
+            frDb.createEntity(newResponse);
+            frDb.deleteEntity(oldResponse);
+        } catch (EntityAlreadyExistsException e) {
+            log.warning("Trying to update an existing response to one that already exists.");
+            throw new EntityAlreadyExistsException(Const.StatusMessages.FEEDBACK_RESPONSE_RECIPIENT_ALREADY_EXISTS);
+        }
+    }
+
+    
+    private void copyOldResponseToNewResponse(FeedbackResponseAttributes newResponse,
+                                    FeedbackResponseAttributes oldResponse) {
         // Copy values that cannot be changed to defensively avoid invalid
         // parameters.
         newResponse.courseId = oldResponse.courseId;
@@ -443,23 +509,8 @@ public class FeedbackResponsesLogic {
         if (newResponse.recipientSection == null) {
             newResponse.recipientSection = oldResponse.recipientSection;
         }
-
-        if (!newResponse.recipientEmail.equals(oldResponse.recipientEmail) ||
-                !newResponse.giverEmail.equals(oldResponse.giverEmail)) {
-            // Recreate response to prevent possible future id conflict.
-            try {
-                newResponse.setId(null);
-                frDb.createEntity(newResponse);
-                frDb.deleteEntity(oldResponse);
-            } catch (EntityAlreadyExistsException e) {
-                log.warning("Trying to update an existing response to one that already exists.");
-                throw new EntityAlreadyExistsException(Const.StatusMessages.FEEDBACK_RESPONSE_RECIPIENT_ALREADY_EXISTS);
-            }
-        } else {
-            frDb.updateFeedbackResponseOptimized(newResponse, oldResponseEntity);
-        }
     }
-
+    
     /**
      * Updates responses for a student when his team changes. This is done by
      * deleting responses that are no longer relevant to him in his new team.

--- a/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java
@@ -442,6 +442,9 @@ public class FeedbackResponsesLogic {
      * in order to prevent an id clash if the previous email is reused later on.
      * @param updatedResponse 
      * @param oldResponseEntity  a FeedbackResponse retrieved from the database 
+     * @throws EntityAlreadyExistsException  if trying to prevent an id clash by recreating a response,
+     *                                       a response with the same id already exist. 
+     * @throws InvalidParametersException               
      */
     public void updateFeedbackResponse(
                         FeedbackResponseAttributes updatedResponse,

--- a/src/main/java/teammates/storage/api/EntitiesDb.java
+++ b/src/main/java/teammates/storage/api/EntitiesDb.java
@@ -61,11 +61,12 @@ public abstract class EntitiesDb {
         
         // TODO: Do we really need special identifiers? Can just use ToString()? 
         // Answer: Yes. We can use toString.
-        if (getEntity(entityToAdd) != null) {
+        Object existingEntity = getEntity(entityToAdd);
+        if (existingEntity != null) {
             String error = String.format(ERROR_CREATE_ENTITY_ALREADY_EXISTS, entityToAdd.getEntityTypeAsString())
                     + entityToAdd.getIdentificationString();
             log.info(error);
-            throw new EntityAlreadyExistsException(error);
+            throw new EntityAlreadyExistsException(error, existingEntity);
         }
         
         Object entity = entityToAdd.toEntity();

